aesd-char-driver/main.c

```diff
@@ -49,6 +50,102 @@ int aesd_release(struct inode *inode, struct file *filp)
     return 0;
 }
 
+//custom llseek function using fixed_size_llseek (Option 2)
+loff_t aesd_llseek( struct file *filp, loff_t offset, int whence )
+{
+    loff_t ret_offset;
+    struct aesd_dev *dev;
+    dev = filp->private_data;
+
+    // locking for the llseek
+    if (mutex_lock_interruptible(&aesd_device.lock)) {
+        ret_offset = -ERESTARTSYS;
+        goto end;
+    }
+
+    ret_offset = fixed_size_llseek(filp, offset, whence, dev->buffer.total_buff_size);
+
+    mutex_unlock(&aesd_device.lock);
+
+    end : return ret_offset;
+}
+
+// function to convert write_cmd and write_cmd_offset into a raw byte offset.
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+    long ret_val = 0;
+    int i;
+    long pos = 0;
+    struct aesd_dev *dev = filp->private_data;
+
+    // check for valid write_cmd
+    if(write_cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+    {
+        ret_val = -EINVAL;
+        goto end;
+    }
+
+    // check for valid write_cmd_offset
+    if(write_cmd_offset > dev->buffer.entry[write_cmd].size)
```

Comment:
@MaanasMDK note that the status here may change between the check and the locking interval below.  You should really be checking this within the locking interval.